components/NewRegistrationForm/NewRegistrationForm.js

```diff
@@ -0,0 +1,104 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { login } from 'common/utils/auth-utils';
+import { insertIf } from 'common/utils/array-utils';
+import MultiStepForm from 'components/Form/MultiStepForm';
+import {
+  InitialStep,
+  PersonalDetailsStep,
+  ClassificationStep,
+  MilitaryStep,
+  TechnologyStep,
+} from './Steps';
+
+class NewRegistrationForm extends Component {
+  static propTypes = {
+    initialValues: PropTypes.objectOf(PropTypes.string),
+  };
+
+  static defaultProps = {
+    initialValues: {
+      // order doesn't matter here
+      ...InitialStep.initialValues,
+      ...PersonalDetailsStep.initialValues,
+      ...ClassificationStep.initialValues,
+      ...MilitaryStep.initialValues,
+      ...TechnologyStep.initialValues,
+    },
+  };
+
+  state = {
+    shouldShowMilitaryStep: false,
+  };
+
+  requireMilitaryStep = () => {
+    this.setState({ shouldShowMilitaryStep: true });
+  };
+
+  hideMilitaryStep = () => {
+    this.setState({ shouldShowMilitaryStep: false });
+  };
+
+  // This method was only created to modularize logic
+  defineSteps = () => {
```

Comment:
Are these objects really necessary? I think you could just use the Components themselves since validationSchema and submitHandler are both static on all of the steps. Something like:

getSteps = () => [
    InitialStep,
    PersonalDetailsStep,
    ClassificationStep,
    // eslint-disable-next-line react/destructuring-assignment
    ...insertIf(this.state.shouldShowMilitaryStep, MilitaryStep),
    TechnologyStep,
  ];

  render() {
    const { initialValues } = this.props;

    return (
      <MultiStepForm
        initialValues={initialValues}
        onFinalStepSuccess={login}
        steps={this.getSteps()}
      />
    );
  }
And in MultiStepForm (note last line):

  render() {
    const { initialValues, steps } = this.props;
    const { errorMessage, stepNumber } = this.state;

    const CurrentStep = steps[stepNumber];
    const isFirstStep = stepNumber === 0;

    return (
      <Formik
        initialValues={initialValues}
        validationSchema={CurrentStep.validationSchema}
        onSubmit={this.handleSubmit}
        render={formikBag => (
          <Form className={styles.MultiStepForm} onSubmit={formikBag.handleSubmit}>
            {/*{CurrentStep.render(formikBag)}*/}
            <CurrentStep {...formikBag} />
           .....

Reply:
This can definitely work.

My reasoning for making `steps` an array of objects is purely to be transparent about the "API" expected of every step. By doing this data structure, we can use proptypes to be more descriptive to any future dev coming to touch multi-step forms.

Reply:
Might be able to achieve the same result if we get fancy with custom propTypes? I was going to play around with some ideas later today, I'm not sure how complicated/unintuitive describing the shape of children components is.

Reply:
Well, no the issue is that static properties wont register as proptypes.

If you mean to imply that we could dive deeper into the proptype of the passed component to somehow determine static properties... wooooo... idek if that's possible

Reply:
What are your thoughts on this?

```

/* eslint-disable unicorn/prevent-abbreviations */
function validStep(propValue, key, componentName, location, propFullName) {
  if (!Object.getOwnPropertyNames(propValue[key]).includes('validationSchema')) {
    return new Error(
      `${propFullName} does not have a validateSchema function.  
      All form steps must contain both validateSchema and submitHandler functions`,
    );
  }

  if (!Object.getOwnPropertyNames(propValue[key]).includes('submitHandler')) {
    return new Error(
      `${propFullName} does not have a submitHandler function.  
      All form steps must contain both validateSchema and submitHandler functions`,
    );
  }
  return null;
}

class MultiStepForm extends React.Component {
  static propTypes = {
    // initialValues must be object where entire form's shape is described
    initialValues: PropTypes.object.isRequired,
    onFinalStepSuccess: PropTypes.func.isRequired,
    steps: PropTypes.arrayOf(validStep).isRequired,
  };
  .....
```

![image](https://user-images.githubusercontent.com/27715246/57329447-65d23580-70d9-11e9-8341-4dcc5e303592.png)

Reply:
I think it's amazing and would like for you to immediately submit a PR :D hahaha